components/places/src/history_sync/payload_evolution_tests.rs

```diff
@@ -0,0 +1,336 @@
+                    "type": 1,
+                },
+            ]
+        })],
+        local_visits: vec![visit("https://dogs.com/", timestamp(10))],
+        incoming_remote_records: vec![json!({
+            "id": "a___________",
+            "title": "dogs",
+            "histUri": "https://dogs.com/",
+            "visits": [],
+            "theme": "nord",
+            // Note: no prank present, so there shouldn't be one in the outgoing unknown_fields
+            "visits": [
+                {
+                    "date": timestamp(1),
+                    "type": 1,
+                },
+                {
+                    "date": timestamp(2),
+                    "type": 1,
+                },
+            ]
+        })],
+        outgoing_unknown_fields: vec![OutgoingRecordUnknownFields {
+            guid: "a___________".into(),
+            fields: unknown_fields!({
+                "theme": "nord",
+            }),
+            visits_fields: vec![
+                (timestamp(1), unknown_fields!({})),
+                (timestamp(2), unknown_fields!({})),
+                (timestamp(10), unknown_fields!({})),
+            ],
+        }],
+    }
+    .run()
+}
+
+#[test]
+fn test_history_visits_roundtrip_fields() {
+    // Test that we roundtrip unknown fields from remote visits after making changes to them
+    RoundtripTest {
+        initial_remote_records: vec![json!({
+            "id": "a___________",
+            "title": "dogs",
+            "histUri": "https://dogs.com/",
+            "visits": [],
+            "visits": [
+                {
+                    "date": timestamp(1),
+                    "type": 1,
+                    // Some more fake new features for the tests:
+                    //   - What emotions did the page evoke?
+                    //   - How much fake news was listed?
+                    "emotion": "joy",
+                    "fake-news-amount": "some",
+                },
+                {
+                    "date": timestamp(2),
+                    "type": 1,
+                    "emotion": "anger",
+                },
+            ]
+        })],
+        local_visits: vec![visit("https://dogs.com/", timestamp(10))],
+        incoming_remote_records: vec![],
+        outgoing_unknown_fields: vec![OutgoingRecordUnknownFields {
+            guid: "a___________".into(),
+            fields: unknown_fields!({}),
+            visits_fields: vec![
+                (
+                    timestamp(1),
+                    unknown_fields!({
+                        "emotion": "joy",
+                        "fake-news-amount": "some",
+                    }),
+                ),
+                (
+                    timestamp(2),
+                    unknown_fields!({
+                        "emotion": "anger",
+                    }),
+                ),
+                (timestamp(10), unknown_fields!({})),
+            ],
+        }],
+    }
+    .run()
+}
+
```

Comment:
Could you please add a test with deduping involved? Maybe one where we have a local record and 2 remote records, all with the same URL but different GUIDs?

AFAICR, history deduping works a little oddly in this case. On the server, only the GUID can be unique, but locally, there's an additional requirement that URLs are unique. But because of how we rotate history, and sync only a subset of your history on the first sync, it's possible for the server to end up with multiple records—each with different GUIDs—for the same URL.

(This is another reason #5472 is tricky: the tombstone only carries the GUID, and if your device deduped the page to a different GUID than the one that uploaded the tombstone, your device won't know to delete that page).

I think our deduping logic just smushes visits for all those together, and picks the first (or the last?) page GUID that it saw as the winner. I'm not sure the right semantics for unknown fields in that case—keep the fields from the record with the winning GUID, and discard the others?—but it would be good to capture them in a test, if it's not too cumbersome to add.

Reply:
> I think our deduping logic just smushes visits for all those together, and picks the first (or the last?) page GUID that it saw as the winner. I'm not sure the right semantics for unknown fields in that case—keep the fields from the record with the winning GUID, and discard the others?—but it would be good to capture them in a test, if it's not too cumbersome to add.

AFAICT, the current behavior is very weird:
 - We pick the GUID from the first record
 - We pick the title from the last record
 - We hope the records are in the same order next sync (which I don't think is guaranteed).

I added a test to verify the title comes from the last record.  Based on that, I made the `unknown_fields` also come from the last record.  I'm not really sure if this is right or not, what do you think?

Reply:
> I added[ a test to verify the title comes from the last record](https://github.com/mozilla/application-services/blob/6879a0b13b62ca23f223130871b0973d51bf8f95/components/places/src/history_sync/plan.rs#L480-L538). Based on that, and the fact that it was simpler to code, I made the `unknown_fields` also come from the last record. I'm not really sure if this is right or not, what do you think?

That sounds great to me, thank you!